test/e2e/operator_test.go

```diff
@@ -4179,3 +4216,63 @@ u3YLAbyW/lHhOCiZu2iAI8AbmXem9lW6Tr7p/97s0w==
 
 	return secret, cl.Create(context.TODO(), secret)
 }
+
+// TestReconcileInternalService verifies that the operator reverts changes to
+// the router-internal service.
+func TestReconcileInternalService(t *testing.T) {
+	t.Parallel()
+	icName := types.NamespacedName{
+		Namespace: operatorNamespace,
+		Name:      "reconcile-internal-service",
+	}
+	domain := icName.Name + "." + dnsConfig.Spec.BaseDomain
+	ic := newPrivateController(icName, domain)
+	if err := kclient.Create(context.TODO(), ic); err != nil {
+		t.Fatal(err)
+	}
+	t.Cleanup(func() { assertIngressControllerDeleted(t, kclient, ic) })
+
+	if err := waitForIngressControllerCondition(t, kclient, 5*time.Minute, icName, availableConditionsForPrivateIngressController...); err != nil {
+		t.Fatalf("failed to observe expected conditions: %v", err)
+	}
+
+	serviceName := controller.InternalIngressControllerServiceName(ic)
+	service := &corev1.Service{}
+	if err := kclient.Get(context.TODO(), serviceName, service); err != nil {
+		t.Fatal(err)
+	}
+
+	findPort := func(name string, service *corev1.Service) *corev1.ServicePort {
+		for i := range service.Spec.Ports {
+			if service.Spec.Ports[i].Name == name {
+				return &service.Spec.Ports[i]
+			}
+		}
+		return nil
+	}
+
+	metricsPort := findPort("metrics", service)
+	if metricsPort == nil {
+		t.Fatalf(`"metrics" port was not found among the service's ports: %v`, service.Spec.Ports)
+	}
+	originalMetricsPortTargetPort := metricsPort.TargetPort
+	metricsPort.TargetPort = intstr.FromInt(12345)
+	if err := kclient.Update(context.TODO(), service); err != nil {
+		t.Fatal(err)
+	}
+
+	if err := wait.PollImmediate(1*time.Second, time.Minute, func() (bool, error) {
+		if err := kclient.Get(context.TODO(), serviceName, service); err != nil {
+			t.Log(err)
+			return false, nil
+		}
+		if metricsPort := findPort("metrics", service); metricsPort == nil {
+			t.Fatalf(`"metrics" port was removed from the service's ports: %v`, service.Spec.Ports)
+		} else if metricsPort.TargetPort == originalMetricsPortTargetPort {
+			return true, nil
+		}
+		return false, nil
```

Comment:
nit: Instead of nil, return an error here that indicates the target port changed when it wasn't expected to?

Reply:
The main reason to use a polling loop here is because the test is verifying that the operator reverts the update: https://github.com/openshift/cluster-ingress-operator/blob/add17d757c103eab60879424b60b42a9448cd401/test/e2e/operator_test.go#L4222-L4224

If the target port is set to `originalMetricsPortTargetPort`, then the operator presumably has reverted the update; else the test most likely needs to wait a little longer: https://github.com/openshift/cluster-ingress-operator/blob/add17d757c103eab60879424b60b42a9448cd401/test/e2e/operator_test.go#L4235-L4238

Returning a non-nil error here would make the test flaky; if the polling loop gets the updated service before the operator manages to revert the update, we should retry, not return an error.